### PR TITLE
apiutil/middleware: add retry logic for obtaining PD leader in redirector

### LINKED ISSUE #8142: add index or lightning or import into failed with error “Error 1105 (HY000): request pd http api failed with status: '503 Service Unavailable” when kill pd leader
### What did you do?
1、add index
2、kill pd leader

### What did you expect to see?
add index can success

### What did you see instead?
add index failed with error “Error 1105 (HY000): request pd http api failed with status: '503 Service Unavailable” 

### What version of PD are you using (`pd-server -V`)?
./pd-server -V
 Release Version: v8.2.0-alpha
Edition: Community
Git Commit Hash: 1679dbca25b3483d1375c7e747da27e99ad77360
Git Branch: heads/refs/tags/v8.2.0-alpha
UTC Build Time:  2024-04-30 11:39:01

### PR DESCRIPTION
**Pull request by JmPotato**
### What problem does this PR solve?

Issue Number: close #8142.

### What is changed and how does it work?

```commit-message
Add retry logic to improve PD HTTP request forwarding success rate during PD leader switch.
```

### Check List

Tests

- Unit test
- Integration test

### Release note

```release-note
None.
```

### CODE REVIEW

**Review by rleungx — APPROVED** [review]
LGTM, but I was thinking that compared with retry in pd, will it be more reasonable to send a request to the leader directly and backoff on the client side?

**Comment by JmPotato** [comment]
> LGTM, but I was thinking that compared with retry in pd, will it be more reasonable to send a request to the leader directly and backoff on the client side?

I agree with this point; client retries should be the most straightforward choice. However, considering that this PR only takes effect during the interval of PD leader re-election, a small internal retry attempt at this time would be lighter than having the client receive an error and then retry. The two are not in conflict.

File: pkg/utils/apiutil/serverapi/middleware.go [hunk 1]
```diff
@@ -233,3 +234,38 @@ func (h *redirector) ServeHTTP(w http.ResponseWriter, r *http.Request, next http
 	client := h.s.GetHTTPClient()
 	apiutil.NewCustomReverseProxies(client, urls).ServeHTTP(w, r)
 }
+
+const (
+	backoffMaxDelay = 3 * time.Second
```
Comment (HuSharp) [inline pkg/utils/apiutil/serverapi/middleware.go:239]: LGTM! a nitpick :) 

Do we need to consider changing this time if the client sets a timeout for the request?
such as `&http.Client{Timeout: 2 * time.Second}`

**Review by HuSharp — APPROVED** [review]

**Comment by lhy1024** [comment]
Does it need to be picked to other branches?

**Comment by JmPotato** [comment]
> Does it need to be picked to other branches?

Let's observe the effect of this change for a period of time before cherry-picking.

**Pull request merged**